### PR TITLE
feat: add fixed-capacity, statically-allocated type tflite::StaticVector

### PR DESCRIPTION
**Pull request by rkuester**
Add a type, tflite::StaticVector, which behaves like std::vector, but
which avoids heap memory allocation.

BUG=#2636

### CODE REVIEW

**Comment by rkuester** [comment]
To my knowledge, we don't have any type like this yet—something that behaves like a std::vector, but can be statically allocated. Am I wrong?

This is used in upcoming PRs.

File: tensorflow/lite/micro/static_vector.h [hunk 1]
```diff
@@ -0,0 +1,90 @@
+// Copyright 2023 The TensorFlow Authors. All Rights Reserved.
```
Comment (rascani) [inline tensorflow/lite/micro/static_vector.h:1]: 2024

File: tensorflow/lite/micro/static_vector.h [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef TENSORFLOW_LITE_MICRO_STATIC_VECTOR_H_
+#define TENSORFLOW_LITE_MICRO_STATIC_VECTOR_H_
+
+#include <array>
+#include <cassert>
+#include <cstddef>
+#include <ostream>
```
Comment (rascani) [inline tensorflow/lite/micro/static_vector.h:21]: Can we avoid ostream usage? I have found just the inclusion of the i/o stream headers results in binary bloat.

**Review by rascani — COMMENTED** [review]
Generally LGTM, just a couple minor comments.

File: tensorflow/lite/micro/static_vector.h (hunk 2 above)
Comment (rkuester) [reply tensorflow/lite/micro/static_vector.h:21]: Sure, that's only there for the operator<<() supplied as a convenience at the bottom. I'll axe it, and either put that in a separate header or just inline something in the one place that used operator<<().

(That lightweight operator<<() is a little perilous anyhow; it doesn't handle all the possible side-effects that using other operator<<()s within its implementation can have on the ostream state.)

**Review by rascani — APPROVED** [review]

**Comment by rkuester** [comment]
These changes required #2648.

**Review by rascani — APPROVED** [review]

**Pull request merged**